### PR TITLE
strict check of pr-review write permission

### PR DESCRIPTION
**Pull request by shogo82148**
There is currently no way to check the available scope of the tokens provided by GitHub; the tokens available for GitHub Actions vary in permissions in complex ways, depending on the repository, workflow settings, and how the workflow is launched.

The most reliable way is to actually post review comments.

- [x] Updated Unreleased section in [CHANGELOG](https://github.com/reviewdog/reviewdog/blob/master/CHANGELOG.md) or it's not notable changes.

close https://github.com/reviewdog/reviewdog/issues/1210

### CODE REVIEW

File: service/github/github.go [hunk 1]
```diff
@@ -369,3 +422,7 @@ func max(x, y int32) int32 {
 	}
 	return x
 }
+
+func RevertMe() {
```
Comment (review-dog) [inline service/github/github.go:426]: **[golint-circleci]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub><br>exported function RevertMe should have comment or be unexported

**Comment by haya14busa** [comment]
Is this pr review ready?

**Comment by shogo82148** [comment]
No, I doesn't work expectly...

**Comment by shogo82148** [comment]
@haya14busa it's ready, can you review it please.

File: service/github/github.go [hunk 2]
```diff
@@ -112,28 +139,57 @@ func (g *PullRequest) postAsReviewComment(ctx context.Context) error {
...
+	// send review comments to GitHub.
 	review := &github.PullRequestReviewRequest{
 		CommitID: &g.sha,
 		Event:    github.String("COMMENT"),
 		Comments: comments,
 		Body:     github.String(g.remainingCommentsSummary(remaining)),
 	}
 	_, _, err := g.cli.PullRequests.CreateReview(ctx, g.owner, g.repo, g.pr, review)
+	if err != nil {
+		log.Println(err)
```
Comment (haya14busa) [inline service/github/github.go:166]: Can we move this line into the isClient error block and use consistent log format?

**Review by haya14busa — CHANGES_REQUESTED** [review]
I'm really sorry for the late review 🙇🙇🙇
It looks promising. Thank you for your work!
I left a few comments.

File: service/github/github.go [hunk 3]
```diff
@@ -28,11 +30,16 @@ const (
 	invalidSuggestionPost = "</details>"
 )
 
+func isClientError(err error) bool {
```
[the remark this answers is not in the document]
Comment (shogo82148) [reply service/github/github.go]: fixed https://github.com/reviewdog/reviewdog/pull/1521/commits/c9884cdaf772bf09479f4bfd20f434935a54a12b

File: service/github/github.go [hunk 4]
```diff
@@ -90,19 +108,28 @@ func (g *PullRequest) Flush(ctx context.Context) error {
...
 	comments := make([]*github.DraftReviewComment, 0, len(g.postComments))
 	remaining := make([]*reviewdog.Comment, 0)
 	for _, c := range g.postComments {
 		if !c.Result.InDiffContext {
 			// GitHub Review API cannot report results outside diff. If it's running
-			// in GitHub Actions, fallback to GitHub Actions log as report .
+			// in GitHub Actions, fallback to GitHub Actions log as report.
 			if cienv.IsInGitHubAction() {
-				githubutils.ReportAsGitHubActionsLog(c.ToolName, "warning", c.Result.Diagnostic)
+				if err := g.logWriter.Post(ctx, c); err != nil {
```
[the remark this answers is not in the document]
Comment (shogo82148) [reply service/github/github.go:138]: fixed in https://github.com/reviewdog/reviewdog/pull/1521/commits/c8c4a58d0ca3263cbcda6681a0719810a4fef0f4

File: service/github/github_test.go [hunk 5]
```diff
@@ -1360,3 +1360,66 @@ func TestGitHubPullRequest_Diff_fake(t *testing.T) {
...
+						Range: &rdf.Range{Start: &rdf.Position{
+							Line: 1,
+						}},
+					},
+					Message: "comment",
+				},
+				InDiffFile:    true,
+				InDiffContext: true,
+			},
+			ToolName: "tool",
```
Comment (haya14busa) [inline service/github/github_test.go:1408]: Can we use a bit more helpful tool name and message?

e.g. tool name: service/github/github_test.go
message:

File: service/github/github_test.go [hunk 6]
```diff
@@ -1360,3 +1360,66 @@ func TestGitHubPullRequest_Diff_fake(t *testing.T) {
...
+	g, err := NewGitHubPullRequest(cli, "o", "r", 14, "sha", "warning")
+	if err != nil {
+		t.Fatal(err)
+	}
+	comments := []*reviewdog.Comment{
+		{
+			Result: &filter.FilteredDiagnostic{
+				Diagnostic: &rdf.Diagnostic{
+					Location: &rdf.Location{
+						Path: "reviewdog.go",
```
Comment (haya14busa) [inline service/github/github_test.go:1398]: Can we use a file under _testdata/ file or service/github/github_test.go if this test expect outputting the actual github action log comments?

File: service/github/github.go [hunk 7]
```diff
@@ -90,29 +113,46 @@ func (g *PullRequest) Flush(ctx context.Context) error {
...
 		if g.postedcs.IsPosted(c, githubCommentLine(c), body) {
+			// it's already posted. skip it.
 			continue
 		}
+
+		rawComments = append(rawComments, c)
 		if !c.Result.InDiffContext {
 			// If the result is outside of diff context, fallback to GitHub Review
 			// Comment API.
 			comment := buildPullRequestComment(c, body, g.sha)
```
Comment (haya14busa) [inline service/github/github.go:154]: [Unrelated to this pr]
I noticed that line (position) information is completely lost for these file comments and it must be confusing for most cases. 🤔
Probably we should add line (position) information in comment body. 

(You don't need to address this comment in this pr. )

File: service/github/github_test.go (hunk 5 above)
Comment (shogo82148) [reply service/github/github_test.go]: fixed https://github.com/reviewdog/reviewdog/pull/1521/commits/0aa828ddf9ed1f1fcc4df51af9841c4a4a201d6d

**Comment by furushchev** [comment]
@haya14busa I'm glad the alternative way was found and merged. (Though, it's a bit sad that the PR #1406 was closed without any interaction after a long pending time)
When will it be released?

**Pull request merged**